proof/hash_to_g1.saw

```diff
@@ -0,0 +1,490 @@
+    };
+
+sgn0x_pty_mont_384_ov <- admit "sgn0x_pty_mont_384" sgn0x_pty_mont_384_spec;
+
+// Proof copied from the proof of `recip_sqrt_fp` in `exp.saw`:
+recip_sqrt_fp_3mod4_ov <- custom_verify "recip_sqrt_fp_3mod4"
+                    ovs_for_reciprocal
+                    recip_sqrt_fp_3mod4_spec
+                    do {
+                      simplify (addsimp recip_sqrt_fp_algebra_thm empty_ss);
+                      unfolding ["sqr_n_mul"];
+                      simplify (addsimps [is_square_fp_thm, sqrt_fp_alg_thm] empty_ss);
+                      simplify fp_exp_simps;
+                      w4_unint_z3 (concat ["fp_exp"] fp_unints);
+                      };
+
+let simp_then_admit r = admit_cryptol (rewrite (cryptol_ss ()) r);
+
+// some rewrites for printing things readably:
+
+let {{
+  fp_mul = Fp.mul
+  fp_sq = Fp.sq
+  fp_add = Fp.add
+  fp_div = Fp.div
+  fp_zero = Fp.field_zero
+  fp_unit = Fp.field_unit
+  fp_is_equal = Fp.is_equal
+  fp_neg = Fp.neg
+}};
+
+readability_rewrites <- for [
+  {{ \p q -> Fp.mul (p,q) == fp_mul (p,q) }}
+, {{ \p q -> Fp.add (p,q) == fp_add (p,q) }}
+, {{ \p -> Fp.sq p == fp_sq p }}
+, {{ \p -> Fp.neg p == fp_neg p }}
+, {{ \p q -> Fp.div (p, q) == fp_div (p, q) }}
+, {{ Fp.field_zero == fp_zero }}
+, {{ Fp.field_unit == fp_unit }}
+, {{ \x y -> Fp.is_equal (x,y) == fp_is_equal (x,y) }}
+  ] simp_then_admit;
+
+let print_readably =  do {
+    simplify (addsimps readability_rewrites empty_ss); // NOTE this modifies the goal
+    print_goal;
+};
+
+// rewrite rules to get rid of the Aux version of some definitions
+
+aux_rewrites <- for [
+    {{ Aux::A' == A' }}
+  , {{ Aux::Z == Z }}
+  , {{ Aux::B' == B' }}
+  , {{ \x -> Aux::sgn0 x == sgn0 x}}
+  , {{ \x -> Aux::sqrt x == sqrt x}}
+] simp_then_admit;
+
+// Assumptions we will not prove:
+
+// We assume that the optimized implementation of the SWU map given in the IETF spec is correct:
+swu_3mod4_correct <- admit_cryptol
+  {{ \u -> ((Fp.div(xn,xd), Fp.div(yn,yd)) where (xn, xd, yn, yd) = Aux::map_to_curve_simple_swu_3mod4 u)
+        == map_to_curve_simple_swu u
+  }};
+
+// The SWU map maps to the curve E'
+swu_on_curve <- simp_then_admit {{ \u -> is_point_affine Curve_E' (map_to_curve_simple_swu u) == True }};
+
+// point_dadd implements add:
+dadd_thm <- admit_cryptol (rewrite (cryptol_ss ()) {{ \ p1 p2 -> affinify Curve_E' (point_dadd Fp p1 p2 A' False) == add Curve_E' (affinify Curve_E' p1) (affinify Curve_E' p2) }});
+
+// the addition of two points on the curve is on the curve:
+point_dadd_on_curve <- simp_then_admit {{ \p q -> is_point_projective Curve_E' (point_dadd Fp p q A' False) == if is_point_projective Curve_E' p /\ is_point_projective Curve_E' q then True else apply is_point_projective Curve_E' (point_dadd Fp p q A' False) }};
+
+// adding a point to its double can be done with POINTonE1_add (cheaper than dadd) because the double of a point is never equal to the point unless the point is point_O:
+// TODO: is this rule okay?
```

Comment:
This looks OK.  I can prove all the cases using the assumptions in `ec_mult.saw`, so it is no worse than those.